lib/spack/spack/compilers/__init__.py

```diff
@@ -134,90 +204,109 @@ def _init_compiler_config(*, scope):
-        scope: configuration scope to modify.
+    If scope is None, all the scopes are searched for removal.
+
+    Args:
+        compiler_spec: compiler to be removed
+        scope: configuration scope to modify
+
+    Returns:
+        True if any compiler has been removed, False otherwise
     """
-    # Need a better way for this
-    global _cache_config_file
+    if not isinstance(compiler_spec, spack.spec.CompilerSpec):
+        compiler_spec = spack.spec.CompilerSpec(compiler_spec)
 
-    compiler_config = get_compiler_config(scope)
-    config_length = len(compiler_config)
+    candidate_scopes = [scope]
+    if scope is None:
+        candidate_scopes = list(spack.config.config.scopes.keys())
 
+    removal_happened = False
+    for current_scope in candidate_scopes:
+        removal_happened |= _remove_compiler_from_scope(compiler_spec, scope=current_scope)
+
+    return removal_happened
+
+
+def _remove_compiler_from_scope(compiler_spec, scope):
+    """Removes a compiler from a specific configuration scope.
+
+    Args:
+        compiler_spec: compiler to be removed
+        scope: configuration scope under consideration
+
+    Returns:
+         True if one or more compiler entries were actually removed, False otherwise
+    """
+    assert scope is not None, "a specific scope is needed when calling this function"
+    compiler_cfg = compiler_config(scope=scope)
     filtered_compiler_config = [
-        comp
-        for comp in compiler_config
+        compiler_entry
+        for compiler_entry in compiler_cfg
         if not spack.spec.parse_with_version_concrete(
-            comp["compiler"]["spec"], compiler=True
+            compiler_entry["compiler"]["spec"], compiler=True
         ).satisfies(compiler_spec)
     ]
 
-    # Update the cache for changes
-    _cache_config_file = filtered_compiler_config
-    if len(filtered_compiler_config) == config_length:  # No items removed
-        CompilerSpecInsufficientlySpecificError(compiler_spec)
-    spack.config.set("compilers", filtered_compiler_config, scope=scope)
+    if len(filtered_compiler_config) == len(compiler_cfg):
+        return False
 
+    # We need to preserve the YAML type for comments, hence we are copying the
+    # items in the list that has just been retrieved
+    compiler_cfg[:] = filtered_compiler_config
+    spack.config.set("compilers", compiler_cfg, scope=scope)
+    return True
 
-def all_compilers_config(scope=None, init_config=True):
-    """Return a set of specs for all the compiler versions currently
-    available to build with.  These are instances of CompilerSpec.
-    """
-    # Get compilers for this architecture.
-    # Create a cache of the config file so we don't load all the time.
-    global _cache_config_file
-    if not _cache_config_file:
-        _cache_config_file = get_compiler_config(scope, init_config)
-        return _cache_config_file
-    else:
-        return _cache_config_file
```

Comment:
Caching here is completely wrong. We cache the result on the first call, and on the second call we reuse it no matter the arguments.